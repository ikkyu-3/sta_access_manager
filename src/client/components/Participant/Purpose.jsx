// @flow
import type { ParticipantPurposeProps } from "@types";

import * as React from "react";
import { MEET_UP, STUDY, WORK, CIRCLE } from "@constants";
import styles from "./style.css";

function Purpose(props: ParticipantPurposeProps): React.Node {
  const { purpose } = props;
  const { columnPurpose, border, work, study, meetUp, circle } = styles;

  let background: string = "";
  let value: string = "";

  if (purpose === MEET_UP) {
    background = meetUp;
    value = "勉強会";
  }
  if (purpose === STUDY) {
    background = study;
    value = "自習";
  }
  if (purpose === WORK) {
    background = work;
    value = "仕事";
  }
  if (purpose === CIRCLE) {
    background = circle;
    value = "サークル";
  }

  return (
    <div className={`${columnPurpose} ${border} ${background}`}>{value}</div>
  );
}
export default Purpose;
