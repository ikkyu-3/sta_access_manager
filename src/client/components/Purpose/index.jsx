// @flow
import type { PurposeProps } from "@types";

import * as React from "react";
import styles from "./style.css";

class Purpose extends React.Component<PurposeProps> {
  componentDidMount() {
    this.props.checkId();
  }

  componentWillUnmount() {
    this.props.removeId();
  }

  render() {
    const { onPurposeButtonClick } = this.props;
    const { container, meetup, work, study, circle, message } = styles;

    return (
      <div className={container}>
        <button id="meetup" className={meetup} onClick={onPurposeButtonClick}>
          勉強会
        </button>
        <button id="work" className={work} onClick={onPurposeButtonClick}>
          仕事
        </button>
        <button id="study" className={study} onClick={onPurposeButtonClick}>
          自習
        </button>
        <button id="circle" className={circle} onClick={onPurposeButtonClick}>
          サークル
        </button>
        <div className={message}>目的を選択して下さい</div>
      </div>
    );
  }
}
export default Purpose;
