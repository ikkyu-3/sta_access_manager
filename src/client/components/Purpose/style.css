@charset "utf-8";
.container {
  display: grid;
  position: relative;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr;
}

.container > button {
  width: 100%;
  color: gray;
  margin: 0;
  padding: 0;
  font-size: 28px;
  font-weight: bold;
  border: none;
  box-sizing: border-box;
  outline: none;
}

.meetup {
  border-right: 2px solid lightgray !important;
  border-bottom: 2px solid lightgray !important;
  border-radius: 5px 0 0 0;
  background-color: #f8f189;
}

.work {
  border-left: 2px solid lightgray !important;
  border-bottom: 2px solid lightgray !important;
  border-radius: 0 5px 0 0;
  background-color: #bbfeff;
}

.study {
  border-right: 2px solid lightgray !important;
  border-top: 2px solid lightgray !important;
  border-radius: 0 0 0 5px;
  background-color: #bcf3cb;
}

.circle {
  border-left: 2px solid lightgray !important;
  border-top: 2px solid lightgray !important;
  border-radius: 0 0 5px 0;
  background-color: #f4d8f5;
}

.message {
  display: flex;
  position: absolute;
  width: 250px;
  height: 60px;
  left: calc(50% - 125px);
  top: calc(50% - 30px);
  justify-content: center;
  align-items: center;
  color: gray;
  font-size: 20px;
  font-weight: bold;
  background-color: white;
  border: 4px solid lightgray;
  border-radius: 10px;
  box-sizing: border-box;
}
