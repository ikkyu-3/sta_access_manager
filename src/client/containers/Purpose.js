// @flow
import type { State, Dispatch } from "@types";

import { connect } from "react-redux";
import { push } from "react-router-redux";

import { requestStart, requestEnd, removeId } from "@actions";
import Purpose from "@components/Purpose";
import {
  API_POST_ENTRY,
  TOP_PATH,
  COMPLETION_ENTRANCE_PATH,
  ERROR_BAD_REQUEST_PATH,
  ERROR_INTERNAL_SERVER_ERROR_PAHT,
  MEET_UP,
  STUDY,
  WORK,
  CIRCLE,
  UNKNOWN
} from "@constants";
import createApiUrl from "@modules/createApiUrl";
import request from "@modules/request";

const mapStateToProps = (state: State) => ({ state });

const mapDispatchToProps = (dispatch: Dispatch) => ({
  dispatch,
  removeId: () => {
    dispatch(removeId());
  }
});

const getPurpose = (purpose: string): string => {
  switch (purpose) {
    case "meetup":
      return MEET_UP;
    case "study":
      return STUDY;
    case "work":
      return WORK;
    case "circle":
      return CIRCLE;
    default:
      return "";
  }
};

const mergeProps = (stateProps, dispatchProps, ownProps) => ({
  ...dispatchProps,
  ...ownProps,
  checkId: () => {
    if (!/^[0-9a-f]{16}$/.test(stateProps.state.id)) {
      dispatchProps.dispatch(removeId());
      dispatchProps.dispatch(push(TOP_PATH));
    }
  },
  onPurposeButtonClick: async (
    e: SyntheticEvent<HTMLButtonElement>
  ): Promise<void> => {
    if (!(e.target instanceof window.HTMLButtonElement)) return;

    const purpose = getPurpose(e.target.id);
    if (purpose === UNKNOWN) return;

    const { id } = stateProps.state;

    const { dispatch } = dispatchProps;
    const { method, path } = API_POST_ENTRY;
    const url = createApiUrl(path);
    const data = { id, purpose };

    // 通信処理
    try {
      dispatch(requestStart());
      await request({ url, method, data });
      dispatch(requestEnd());

      dispatch(push(COMPLETION_ENTRANCE_PATH));
    } catch (error) {
      dispatch(requestEnd());

      if (error.status === 500) {
        dispatch(push(ERROR_INTERNAL_SERVER_ERROR_PAHT));
      } else {
        dispatch(push(ERROR_BAD_REQUEST_PATH));
      }
    }
  }
});

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(
  Purpose
);
