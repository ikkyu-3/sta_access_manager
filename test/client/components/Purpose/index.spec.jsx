import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Purpose from "@components/Purpose";

describe("Purpose.jsx", () => {
  describe("スナップショット", () => {
    it("正しいレンダリング", () => {
      const fn = () => {};
      const tree = renderer
        .create(
          <Purpose checkId={fn} removeId={fn} onPurposeButtonClick={fn} />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("コンポーネントのテスト", () => {
    const checkIdMock = jest.fn();
    const removeIdMock = jest.fn();
    const onPurposeButtonClickMock = jest.fn();
    let purpose;

    beforeAll(() => {
      purpose = shallow(
        <Purpose
          checkId={checkIdMock}
          removeId={removeIdMock}
          onPurposeButtonClick={onPurposeButtonClickMock}
        />
      );
    });

    beforeEach(() => {
      onPurposeButtonClickMock.mockClear();
    });

    it("componentDidMountは、checkIdを実行する", () => {
      expect(checkIdMock).toBeCalled();
    });

    it("componentWillUnmountは、removeIdを実行する", () => {
      purpose.instance().componentWillUnmount();
      expect(removeIdMock).toBeCalled();
    });

    it("勉強会をクリックしたら、onPurposeButtonClickが実行される", () => {
      purpose.find("#meetup").simulate("click");
      expect(onPurposeButtonClickMock).toBeCalled();
    });

    it("仕事をクリックしたら、onPurposeButtonClickMockが実行される", () => {
      purpose.find("#work").simulate("click");
      expect(onPurposeButtonClickMock).toBeCalled();
    });

    it("自習をクリックしたら、onPurposeButtonClickMockが実行される", () => {
      purpose.find("#study").simulate("click");
      expect(onPurposeButtonClickMock).toBeCalled();
    });

    it("サークルをクリックしたら、onPurposeButtonClickMockが実行される", () => {
      purpose.find("#circle").simulate("click");
      expect(onPurposeButtonClickMock).toBeCalled();
    });
  });
});
