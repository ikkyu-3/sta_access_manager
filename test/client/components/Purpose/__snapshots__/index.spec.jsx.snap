// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Purpose.jsx スナップショット 正しいレンダリング 1`] = `
<div
  className="container"
>
  <button
    className="meetup"
    id="meetup"
    onClick={[Function]}
  >
    勉強会
  </button>
  <button
    className="work"
    id="work"
    onClick={[Function]}
  >
    仕事
  </button>
  <button
    className="study"
    id="study"
    onClick={[Function]}
  >
    自習
  </button>
  <button
    className="circle"
    id="circle"
    onClick={[Function]}
  >
    サークル
  </button>
  <div
    className="message"
  >
    目的を選択して下さい
  </div>
</div>
`;
